cpu/mips32r2_common/Makefile.include

```diff
@@ -10,13 +6,36 @@ export MEMORY_BASE=0x80000000
 export MEMORY_SIZE=1M
 export APP_START=0x80000000
 
-include $(MIPS_ELF_ROOT)/share/mips/rules/mipshal.mk
+# Portable 'lowercase' func.
+lc = $(subst A,a,$(subst B,b,$(subst C,c,$(subst D,d,$(subst E,e,$(subst F,f,$(subst G,g,$(subst H,h,$(subst I,i,$(subst J,j,$(subst K,k,$(subst L,l,$(subst M,m,$(subst N,n,$(subst O,o,$(subst P,p,$(subst Q,q,$(subst R,r,$(subst S,s,$(subst T,t,$(subst U,u,$(subst V,v,$(subst W,w,$(subst X,x,$(subst Y,y,$(subst Z,z,$1))))))))))))))))))))))))))
+
+# Default values for the linker script symbols listed below are
+# defined in the linker script.
+
+# These are linker script symbols that are prefixed with '__"
+priv_symbols = MEMORY_BASE MEMORY_SIZE STACK
+priv_symbols += ENABLE_XPA
+priv_symbols += FLUSH_TO_ZERO
+priv_symbols += FLASH_START APP_START FLASH_APP_START
+priv_symbols += ISR_VEC_SPACE ISR_VECTOR_COUNT
+
+comma := ,
+
+# A bit of makefile magic:
+# foreach symbol in overridable ld-symbols :
+#   If symbol has a value, produce a linker argument for that symbol.
+MIPS_HAL_LDFLAGS = $(foreach a,$(priv_symbols),$(if $($a),-Wl$(comma)--defsym$(comma)__$(call lc,$(a))=$($a)))
 
 # define build specific options
-export CFLAGS_CPU   = -EL -march=mips32r2 -std=gnu99
+# Remove this once the MIPS toolchain headers are updated to include upstream
+# newlib commit 81c17949f0419d1c4fee421c60987ea1149522ae
+# https://cygwin.com/git/gitweb.cgi?p=newlib-cygwin.git;a=commitdiff;h=81c17949f0419d1c4fee421c60987ea1149522ae
+# Otherwise we get an error about a missing declaration of strnlen in some parts.
+export CFLAGS += -std=gnu99
+export CFLAGS_CPU   = -EL -march=mips32r2 -mabi=$(ABI)
 export CFLAGS_LINK  = -ffunction-sections -fdata-sections -fno-builtin -fshort-enums
-export CFLAGS_DBG = -O0 -g2
-export CFLAGS_OPT = -Os -g2
+export CFLAGS_DBG   = -g3
+export CFLAGS_OPT   = -Os
```

Comment:
@neiljay you have misinterpreted the intent of the CFLAGS_DBG and CFLAGS_OPT variables in the other Makefiles, they are simply groupings for making the file easier to read:
 - CFLAGS_DBG contains flags which affect the generation of debug symbols
 - CFLAGS_OPT contains flags which affect the code optimization
If you check the other platforms you'll see that they are both unconditionally added to CFLAGS.
We don't have any concept of Release build vs. Debug build in the current RIOT build system.
I forgot to remove the comment on CFLAGS_DBG below, though.